pages/blog/add-reactivity-to-an-existing-source.mdx

```diff
@@ -0,0 +1,170 @@
+
+Across my researches I found one cool public endpoint regarding [SpaceX](https://api.spacex.land/graphql/). It has `Queries` and `Mutations` ready to use!
+So I decided to pimp this endpoint, and bring a bit of reactivity with subscriptions.
+
+The idea is to publish the new user when it's inserted with this `mutation`:
+
+```graphql
+mutation addUser {
+  insert_users(objects: { name: "jyc" }) {
+    affected_rows
+    returning {
+      name
+    }
+  }
+}
+```
+
+_How hard is to do this? Let's find out..._
+
+> Let's _Pimp this graphQL_!
+
+![Pimp my graphQL](/blog-assets/add-reactivity-to-an-existing-source/PimpMyGraphQL.gif)
+
+# Implementation
+
+_Multiple ways are possible!_
+
+<details>
+  <summary>Spoiler alert</summary>
+  Thanks to `graphql-mesh`, it's dead easy! 🚀🚀🚀
+</details>
+
+1. Create a new repo
+
+```bash
+mkdir newRepo
+cd newRepo
+yarn init
+```
+
+2. Add `graphql-mesh` packages
+
+```bash
+yarn add @graphql-mesh/cli @graphql-mesh/graphql @graphql-mesh/transform-resolvers-composition graphql
+```
+
+3. Add `.meshrc.yaml` file in the root folder with
+
+```yaml
+sources:
+  - name: spacexGQL
+    handler:
+      graphql:
+        endpoint: https://api.spacex.land/graphql/
+
+additionalTypeDefs: |
+  extend type Subscription {
+    usersAdded: [users]
+  }
+
+transforms:
+  - resolversComposition:
+      - resolver: 'Mutation.insert_users'
+        composer: ./src/composition/insert_user
+
+additionalResolvers:
+  - type: Subscription
+    field: usersAdded
+    pubsubTopic: usersAdded
+```
+
+Where I
+
+- declare the graphQL source
+- extend `Subscription` to listen to user added users
+- compose the existing `Mutation.insert_users`
+
+4. Add `./src/composition/insert_user.js` file with all the logic for this new mutation composition.
+
+```js
+module.exports = (next) => async (root, args, context, info) => {
+  // add returning.id to the selection
+  info.operation.selectionSet.selections[0].selectionSet.selections.push({
+    kind: 'Field',
+    name: {
+      kind: 'Name',
+      value: 'returning',
+    },
+    arguments: [],
+    directives: [],
+    selectionSet: {
+      kind: 'SelectionSet',
+      selections: [
+        {
+          kind: 'Field',
+          name: {
+            kind: 'Name',
+            value: 'id',
+          },
+          arguments: [],
+          directives: [],
+        },
+      ],
+    },
+  });
+
+  // send the mutation to the graphQL source
+  const result = await next(root, args, context, info);
+
+  // select returned ids
+  const ids = result.returning.map((c) => c.id);
+
+  // Query users enforcing fields to return
+  const responseUser = await context.spacexGQL.apiQuery.users(
+    {
+      where: { id: { _in: ids } },
+    },
+    {
+      fields: {
+        id: true,
+        name: true,
+        rocket: true,
+        timestamp: true,
+        twitter: true,
+      },
+    }
+  );
+
+  // publish new users
+  context.pubsub.publish('usersAdded', responseUser);
+
+  return result;
```

Comment:
One nitpick I have is that the fields we added on the fly will now also be included in the mutation result, even if we did not select the specific fields. So we should also take care of removing them from the result

Reply:
The all logic should be extracted probably.
Add `id` if not requested & Removing it if it was not requested.